### PR TITLE
Camera API

### LINKED ISSUE #13364: Make 3rd person "camera offset" server controllable
## Problem

Right now this is hardcoded: https://github.com/minetest/minetest/blob/master/src/client/camera.cpp#L434
Some mods that let the player attach to larger entities (like steampunk_blimp) would benefit from being to control this.

2.75 meters is useless in that case. 1/2 the time you are inside the blimp somewhere. Here 20 meters make much more sense.

## Solutions

Add a 3rd-person camera offset API. It allows setting a single. The offset - in camera direction - from the camera.

## Alternatives

One could also relax the setEyeOffset limits (especially in the Z-direction). But note, that that moves the camera, so the player the is not in the center of the view as one pans up or down.

## Additional context

set_eye_offset limits relaxed and set to (0, 20, -100):

screenshot_20230327_142946
screenshot_20230327_142954
screenshot_20230327_142943

With "camera offset" increased by 10:
screenshot_20230327_143037
screenshot_20230327_143041
screenshot_20230327_143047

Unchanged:
screenshot_20230327_143415
screenshot_20230327_143419
screenshot_20230327_143423

### PR DESCRIPTION
**Pull request by kilbith**
### Introduction and credits

Here it is, folks. The Camera API. The one that will dramatically contribute to the modding possibilities.
This API represents months of work of successive API design iterations, and several discouragement / re-motivation cycles. It has been financially supported by the French state as part of a mission with the [Pole Emploi](https://www.pole-emploi.fr/accueil/). I want to thank them to allow me to dedicate all my time to develop this.

@x2048 also has been of **crucial help** privately and nothing would have been possible without him. He is the co-author of this.
I also want to thank the community for their support through my sneak-peek [dev demos](https://www.youtube.com/@jean-patrickguerrero422/videos).

Closes #13364

### How to use

This PR contains a demo mod in `devtest`. I've uploaded a demo video **[HERE](https://www.youtube.com/watch?v=gfb31TnOixw)**.
API methods are also documented in `lua_api.txt`.

The secondary cameras are all created in the client environments, each of them have their own pool of cameras.
The server governs them all.

If you want to manipulate the camera with the mouse cursor (translate, rotate, zoom), you need first to set up a GUI (formspec), like in any RTS game. See devtest mod' code as example.

### Known issues

- [ ] **The most critical one and not an easy one: the occlusion culling.** You may want to disable it in the code to test the API in the best conditions. A new way of doing the occlusion must be found (like, from the camera view direction and not the player one). @x2048 says it'd need a new draw list per camera and it's pretty expensive. @celeron55 may have another idea?
- [ ]  Position and rotation interpolations must be reworked, especially the rotation one which is not working on X/Y. Low-hanging fruits but I believe other people know better than me how to do it properly.

I'll gladly take care of your PRs against my fork to collectively improve this API. It you want to make it happen in upstream, it only depends **on you**.

### Removed features

In order to make this PR mergeable, I've decided to remove some features that I've shown in my videos.

- The [rectangle selection](https://www.youtube.com/watch?v=W7JPVcauq9c) and the [click-picking](https://www.youtube.com/watch?v=VED5EEcYxTo) of objects and positions.
It involved to modify Irrlicht and bring back thousands of lines of removed code (namely: all the `TriangleSelector` classes which are bloat).
- The controlling of the player's yaw with the keys (i.e. you could control the player with your keyboard only). I felt it wasn't part of a Camera API.

### CODE REVIEW

**Comment by Desour** [comment]
Nice features.
But sadly this will still require much work until it works bug-free.
https://github.com/kilbith/minetest/pull/1

File: src/client/universalcamera.cpp [hunk 1]
```diff
@@ -0,0 +1,449 @@
...
+			m_interp.speed.zoom = 0.f;
+		}
+
+		updatePos(m_rot);
+	}
+
+	if (inter_rot) {
+		if (m_interp.rotation_counter != v3f()) {
+			// TODO: X/Y interpolated rotation is not working correctly (needs reworking)
+			// Z rolling is OK
```
Comment (kilbith) [inline src/client/universalcamera.cpp:315]: Reviewer checkpoint.

File: src/client/universalcamera.cpp [hunk 2]
```diff
@@ -0,0 +1,449 @@
...
+			if (m_interp.rotation_counter.Z > 0.f)
+				m_roll += step;
+		} else {
+			m_target = m_interp.target;
+			m_roll = m_interp.rotation.Z;
+			m_interp.speed.rotate = 0.f;
+		}
+	}
+
+	if (inter_pos) {
```
Comment (kilbith) [inline src/client/universalcamera.cpp:344]: Reviewer checkpoint: position interpolation should use Irrlicht's `getInterpolated()`

**Comment by kilbith** [comment]
Idea from @x2048, FWIW:

> I have an idea for occlusion culling, inspired by portal OC
> 1. For each mapblock mark the walls as transparent/opaque based on the nodes at the edge
> 2. Treat each mapblock as a room at walls as portals and rub the portal OC algorithm
> 3. Because it is a regular grid, the math should be very easy and fast

**Comment by lhofhansl** [comment]
What's needed to finish this?

As for server occlusion culling. There's a (server) flag to disable it. If we wanted to, could make this controllable via Lua, then a mod could disable that if needed.
(Number of blocks sent over the network would significantly increase - depending on the scene 100x or more, but typically maybe 10x, so definitely not recommended on a busy server)

Although here we are also talking client side, right? Would need Lua to be able to disable both, I suppose.

**Comment by kilbith** [comment]
- People need to rework position/rotation interpolations.
- I'm going to commit a fix for the camera offset which is not applied in some cases.
- @x2048 needs to work on draw lists (shadows + map) per camera.
- Some motivated testers to try this.

**Comment by x2048** [comment]
I've identified three areas which are currently hard-wired to the player and will need to be changed:
- Draw lists (mentioned above)
- Shadow mapping - light space frustum, shadow draw list, textures etc.
- Server-side code for sending mapblocks

Here is how I see it:
* Another PR is needed to untie drawlist and shadow draw list assembly from ClientMap / Map, so that we can have many of them. This can be done in another PR before this PR.
* Shadow Mapping must be done per camera, with lower resolution and smaller draw list, maybe even spread render over more frames to minimize impact if the camera is static. - it could be done in this PR, but we need to see yet how much needs to be done.

  At this point we will get playable camera API limited to the player's view range.

* Mapblock-sending code needs to be made aware of multiple location points per player (cameras), so that it can send blocks around cameras but not arbitrary blocks of the map, and cameras can be placed anywhere on the map. I think this can be done in a later PR.

**Comment by x2048** [comment]
> (Number of blocks sent over the network would significantly increase - depending on the scene 100x or more, but typically maybe 10x, so definitely not recommended on a busy server)

This has direct impact on FPS right now, we need changes to mapblock receiving code.

**Comment by Wuzzy2** [comment]
Your demo video already reveals a bug: Like, the shadows of everything change while moving the player.

**Comment by kilbith** [comment]
It's a shadow draw list / occlusion issue. Yes, even shadows are occluded. Big surprise.

**Comment by Wuzzy2** [comment]
More bugs/problems:

* Cameras seem to be singleplayer-centric. **If the player moves too far away from a fixed-position camera, the area the camera watches will not render and is no longer visible in the camera**
* The DevTest commands are poorly documented. The syntax help (`params`) is missing
* DevTest uses `core.*` instead of `minetest.*`
* `player:set_camera(-1, {enabled=false})` does not do anything although the documentation claims that camera -1 can be deactivated
* `player:set_camera(-1, {pos={x=0,y=0,z=0})` does not work either
* In fact, I cannot get anything to work with camera -1
* DevTest: `/set_camera mapnip` and then pressing Esc causes weird behavior. Maybe it should be made clearer what I am supposed to do / what mode I am in?
* Punching the camera entity in DevTest displays an error that `default_camera` image does not exist
* DevTest: If I destroy the camera entity, many `/set_camera` commands stop working
* Documentation of `viewport` is bad (what is a "rectangle of floats"?)
* No error or warning if I use `set_camera` or `remove_camera` in Lua with invalid ID

**Comment by kilbith** [comment]
`-1` is the main camera, not a custom one added from `add_camera`. You are not supposed to configure it with this API.

> DevTest: /set_camera manip and then pressing Esc causes weird behavior. Maybe it should be made clearer what I am supposed to do / what mode I am in?

Mouse manipulation is only possible when a GUI is currently displayed. Pressing Esc closes the GUI, thus removes this ability.

**Comment by Wuzzy2** [comment]
OK, in that case, the documentation is false: “`-1` is the main camera which is also deactivatable.”. Except I cannot deactivate the main camera. What's the point of -1 if you can't do anything with it? I think the purpose of -1 should be explained in more detail in the documentation (from a Lua perspective, that is).

> Mouse manipulation is only possible when a GUI is currently displayed. Pressing Esc closes the GUI, thus removes this ability.

OK, it might make sense to show some dummy GUI instead of an empty one for DevTest then so that you know when you're still "in the formspec". IMHO an empty GUI is too confusing, even for testing, since you can't be sure in which state you're in. Could a very small dummy formspec that just says "press Esc to close" or something similar.

**Comment by lhofhansl** [comment]
Can we do something here before have it perfect with this?
There are many simple things you cannot do today, even something as simple as #13364

I'm afraid we're doing the **Minetest Thing(tm)** here again: Attempt something perfect, which is never completely finished, meanwhile all other simpler partial solutions are blocked and waiting for the one that never finishes.

So, practically, I am asking what of this can we have before we solve the complete topic? Many cameras with a drawlist each will, IMHO, never work, we have performance problems with a single drawlist already.

For example, can we start with an API for the main camera only? Take #13364 as inspiration :)

**Comment by Wuzzy2** [comment]
:-(

Does this PR need adoption? It would be sad to have this code die and all the hard work go to waste.

**Comment by lhofhansl** [comment]
This is too ambitious anyway.
Let's start with being able to control the (one) main camera.

**Comment by tigercoding56** [comment]
> This is too ambitious anyway. Let's start with being able to control the (one) main camera.
ambition is good , else nothing gets done

**Comment by tigercoding56** [comment]
> :-(
> 
> Does this PR need adoption? It would be sad to have this code die and all the hard work go to waste.
it already has gone to waste  ,  the camera-api branch is deleted

**Pull request closed without merging**